### ISSUE TITLE
Quaternion works wrong with bone-based (deformation) animation

### ISSUE THREAD

**Issue by Garrom** [godotengine/godot-blender-exporter#142]
**OS:**
GNU/Linux Ubuntu

**Godot version:**
Beta 2

**Blender version:**
2.79b

**Issue description:**
Exporting to newer versions of Godot have destroyed animations. From merge [#19193](https://github.com/godotengine/godot/pull/19193) quat works little different making this plugin unusable. While playing animation, every frame is ` core/math/quat.cpp:151 - Condition ' !q.is_normalized() ' is true. returned: Quat()` or ` core/math/quat.cpp:150 - Condition ' !is_normalized() ' is true. returned: Quat()
` added to error log. Animations not dealing with rotation and just moving bones without rotating them are unaffected. Only bones are affected, if I simply rotate whole mesh (which is not possible for deformation animations) bug would not occur.

**Minimal reproduction project:**

I did as simple as I could. Behold my cube ! just use plugin to export it.
[Cube.zip](https://github.com/godotengine/godot-blender-exporter/files/2776018/Cube.zip)

**Comment by Garrom** [comment]
@Jason0214  Alright, I took a look at it. Aparently, godot needs quaternion to be normalized (values in range between 0-1). I tried to look trough your code but I'm unable to pinpoint where are animation keyframes.

**Comment by Garrom** [comment]
Sorry, pressed wrong button.

**Comment by Garrom** [comment]
I found problem is that values inside quaternion are not in range between 0 to 1.  I looked on plugin but I'm unable to pinpoint where are you dealing with keyframe rotation. I don't know python but there must be way how to normalize numbers.

**Comment by Jason0214** [comment]
I did some investigation the other day, I am not able to reproduce the issue as you described. But I checked some quaternion frames length square,  some example I got:
```
0.9999999890563667
1.000000003439075
1.0000000174107841
1.0000001384144854
0.999999965187933
0.9999999703071332
1.0000000235420325
1.0000000753559002
0.9999999940913824
0.9999999945307865
1.0000002036630926
1.0000000081613516
0.9999999870227114
1.000000011572027
0.9999999808575879
1.0000000076473228
```
looks like they are all normalized with a reasonable precision.

are you suggesting each value of a quaternion should be in range (0, 1)? that could be a problem, because some of the quaternion has negative `w` value. 

---
by the way the exporter does quaternion writing here https://github.com/godotengine/godot-blender-exporter/blob/master/io_scene_godot/converters/animation/serializer.py#L319

**Comment by Jason0214** [comment]
is it possible to normalize a quaternion to get all the four dimensions to be positive 🤔?     could you give a point to where godot does this range (0, 1) check？

**Comment by Garrom** [comment]
> ? could you give a point to where godot does this range (0, 1) check？
Sure thing ! It is exactly this [commit and this file](https://github.com/godotengine/godot/pull/19193/commits/9d41161596167984efdb02639d0857da8979b01b#diff-301040857fda0be8b8ec8e1bc218e548). Errors I get are created by lines 149 and 150 (150 and 151 in godot, dude, all lines should start by 0) There is `is_normalized()` call which returns true if all values are in range from 0 to 1, otherwise return false.

> is it possible to normalize a quaternion to get all the four dimensions to be positive ?
Yeah, I believe so. If they can make humans fly, everything is possible !

**Comment by Garrom** [comment]
Alright [here](https://math.stackexchange.com/questions/1703466/normalizing-a-quaternion) is some math.

I also found if you put all four values of quaternion into array you can use [this](https://stackoverflow.com/questions/26785354/normalizing-a-list-of-numbers-in-python) . They all speak in python which is language I don't understand but I still want to help.

**Comment by Jason0214** [comment]
But the above 'normalize' is all about normalize the quaternion length to `1`, which I think the exporter already did.   Besides, the commit you are referring to is quite old commit ( even older than the exporter .. )  I am thinking it might still be some precision problem you are hitting

**Comment by Jason0214** [comment]
I am not able to reproduce the issue either on my OSX or ubuntu, so could you give a try #146 ?  I add a force normalizing quaternion, maybe it can solve your issue.

**Comment by Jason0214** [comment]
Ah, I am able to reproduce it with the beta2 version in your link.   (Last time I tested it on a custom build which I pulled the latest master branch. )

Probably this is an bug at Godot side and having been fixed recently.

**Comment by Garrom** [comment]
@Jason0214 Just tried offlicial [beta 6](https://downloads.tuxfamily.org/godotengine/3.1/beta6/) Problem is still here. Animation system was reworked so I belive that is where problem is.

**Comment by Jason0214** [comment]
well, this time able to reproduce on both linux and osx,  I will look into it

**Comment by Jason0214** [comment]
Looks like #146 fixes the issue on OSX, I will find sometime to test it on linux.   Besides, based on the Godot log (quaternion non-normalized), I think it is the correct fix

**Comment by Garrom** [comment]
@Jason0214  I just noticed that if I set keyframe interpolation mode to "nearest" it works just fine (inside godot). Can you please confirm this on linux ? Unfortunately, I'm usually using linear or cubic... Both don't work. Looks like interpolation error which just proving my theory that it is fault of rewritten animation player.

**Comment by Jason0214** [comment]
Yeah, it could be quaternion precision not handled very well in the new implementation. Maybe it could a good idea to open an issue in Godot side and link this thread, and I will try to find some time to talk with developers.

**Comment by Garrom** [comment]
@Jason0214  [Issue opened](https://github.com/godotengine/godot/issues/26363)

**Comment by doyouknowaboutq** [comment]
i'm using blender 2.80 and i tested my model with both of godot 3.0.6 and 3.1 official releases.
here is blender animation with 3 keyframes:
![blenderanim](https://user-images.githubusercontent.com/48880100/55158678-828d5c00-5170-11e9-98a3-0759a2323a74.png)

here is blender export settings:
![blenderexport](https://user-images.githubusercontent.com/48880100/55158686-8de08780-5170-11e9-9e33-fe7539e84075.png)

here is 3.1 animation player:
![31anim](https://user-images.githubusercontent.com/48880100/55158716-9c2ea380-5170-11e9-8fef-45aadcae915a.png)
the animation does not work at linear mode. it only works with nearest mode.
and you know there is a lot of quat errors.

here is 3.0.6 animation player:
![306anim](https://user-images.githubusercontent.com/48880100/55158828-d8fa9a80-5170-11e9-8d43-e317acb3b3fa.png)
there is no quat errors.
the aniamtion working with nearest and linear mode. but why?

why the plugin is exporting animations with too many keyframes like that? (this problem there was in blender 2.79)

**Comment by phaitonican** [comment]
Have the same bug with Blender 2.8 and latest release.
It seems (like above said) adding ANY Object Constraint in Blender to the object solves the problem for me. (First I tried out the rotation limiter then also anything else)

**Comment by Garrom** [comment]
I just found out that XYZ Euler works good. Looks like problem applies only to WXYZ Quaternion. Using XYZ Euler seems to be solution, or at least workaround
